converting json to vector of type with templated constructor

**Bug Report**

The following code fails to compile when uncommenting line 16:
```c++
#include <nlohmann/json.hpp>

namespace nl = nlohmann;

struct S {
  S() = default;
  template<typename T> S(const T& t) {}
};

void from_json(const nl::json& j, S& s) {
}

int main() {
  nl::json j;
  auto s = j.get<S>();
  //auto v = j.get<std::vector<S>>();
};
```
The error I get is 
```
json.hpp(1360): error C2338: could not find from_json() method in T's namespace
```

The cause seems to stem from the check for `std::is_convertible<BasicJsonType, typename CompatibleArrayType::value_type>::value` on `void from_json(const BasicJsonType& j, CompatibleArrayType& arr)`. 

Since you do not convert directly from json to the array value but call `get()` I think this check can be changed to allow this to compile.

Compiled on VS 15.6.0 preview 2.0
Library version 3.0.1

The culprit is indeed the `is_convertible` check, I added the following method (following this [cppreference](http://en.cppreference.com/w/cpp/types/is_convertible) page)

Here is the output:

>t.cpp:14:19: error: conversion from 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator,
      adl_serializer>' to 'S' is ambiguous
T test() { return std::declval<F>(); }
                  ^~~~~~~~~~~~~~~~~
t.cpp:24:3: note: in instantiation of function template specialization 'test<nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long,
      unsigned long long, double, std::allocator, adl_serializer>, S>' requested here
  test<nl::json, S>();
  ^
develop/json.hpp:3053:5: note: candidate function [with ValueType = S, $1 = 0]
    operator ValueType() const
    ^
t.cpp:7:25: note: candidate constructor [with T = nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double,
      std::allocator, adl_serializer>]
  template <typename T> S(const T &t) {}

This ambiguity is what prevent your code from working, I will open a PR to replace the `is_convertible` to `is_constructible`.

Thanks for the report!

Thanks. The actual type is from a library so I can't change the constructor.

I'm not sure is_constructible is appropriate here. I think you just need to
check that get<T>() is valid.

On 19 Jan 2018 12:09 pm, "Théo DELRIEU" <notifications@github.com> wrote:

> The culprit is indeed the is_convertible check, I added the following
> method (following this cppreference
> <http://en.cppreference.com/w/cpp/types/is_convertible> page)
>
> Here is the output:
>
> t.cpp:14:19: error: conversion from 'nlohmann::basic_json<std::map,
> std::vector, std::__1::basic_string, bool, long long, unsigned long long,
> double, std::allocator,
> adl_serializer>' to 'S' is ambiguous
> T test() { return std::declval(); }
> ^~~~~~~~~~~~~~~~~
> t.cpp:24:3: note: in instantiation of function template specialization
> 'test<nlohmann::basic_json<std::map, std::vector, std::__1::basic_string,
> bool, long long,
> unsigned long long, double, std::allocator, adl_serializer>, S>' requested
> here
> test<nl::json, S>();
> ^
> develop/json.hpp:3053:5: note: candidate function [with ValueType = S, $1
> = 0]
> operator ValueType() const
> ^
> t.cpp:7:25: note: candidate constructor [with T =
> nlohmann::basic_json<std::map, std::vector, std::__1::basic_string, bool,
> long long, unsigned long long, double,
> std::allocator, adl_serializer>]
> template S(const T &t) {}
>
> This ambiguity is what prevent your code from working, I will open a PR to
> replace the is_convertible to is_constructible.
>
> Thanks for the report!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nlohmann/json/issues/924#issuecomment-358922173>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHTojR_ePuofeCR6m7uAB3ejqCjb8C74ks5tMGnSgaJpZM4Rjs5L>
> .
>

--- a/include/nlohmann/detail/conversions/from_json.hpp
+++ b/include/nlohmann/detail/conversions/from_json.hpp
@@ -177,15 +177,21 @@ void from_json_array_impl(const BasicJsonType& j, std::array<T, N>& arr, priorit
     }
 }
 
-template<typename BasicJsonType, typename CompatibleArrayType,
-         enable_if_t<is_compatible_array_type<BasicJsonType, CompatibleArrayType>::value and
-                     std::is_convertible<BasicJsonType, typename CompatibleArrayType::value_type>::value and
-                     not std::is_same<typename BasicJsonType::array_t, CompatibleArrayType>::value, int> = 0>
+template <
+    typename BasicJsonType, typename CompatibleArrayType,
+    enable_if_t <
+        is_compatible_array_type<BasicJsonType, CompatibleArrayType>::value and
+        not std::is_same<typename BasicJsonType::array_t,
+                         CompatibleArrayType>::value and
+        std::is_constructible <
+            BasicJsonType, typename CompatibleArrayType::value_type >::value,
+        int > = 0 >
 void from_json(const BasicJsonType& j, CompatibleArrayType& arr)
 {
     if (JSON_UNLIKELY(not j.is_array()))
     {
-        JSON_THROW(type_error::create(302, "type must be array, but is " + std::string(j.type_name())));
+        JSON_THROW(type_error::create(302, "type must be array, but is " +
+                                      std::string(j.type_name())));
     }
 
     from_json_array_impl(j, arr, priority_tag<2> {});

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -1084,15 +1084,21 @@ void from_json_array_impl(const BasicJsonType& j, std::array<T, N>& arr, priorit
     }
 }
 
-template<typename BasicJsonType, typename CompatibleArrayType,
-         enable_if_t<is_compatible_array_type<BasicJsonType, CompatibleArrayType>::value and
-                     std::is_convertible<BasicJsonType, typename CompatibleArrayType::value_type>::value and
-                     not std::is_same<typename BasicJsonType::array_t, CompatibleArrayType>::value, int> = 0>
+template <
+    typename BasicJsonType, typename CompatibleArrayType,
+    enable_if_t <
+        is_compatible_array_type<BasicJsonType, CompatibleArrayType>::value and
+        not std::is_same<typename BasicJsonType::array_t,
+                         CompatibleArrayType>::value and
+        std::is_constructible <
+            BasicJsonType, typename CompatibleArrayType::value_type >::value,
+        int > = 0 >
 void from_json(const BasicJsonType& j, CompatibleArrayType& arr)
 {
     if (JSON_UNLIKELY(not j.is_array()))
     {
-        JSON_THROW(type_error::create(302, "type must be array, but is " + std::string(j.type_name())));
+        JSON_THROW(type_error::create(302, "type must be array, but is " +
+                                      std::string(j.type_name())));
     }
 
     from_json_array_impl(j, arr, priority_tag<2> {});

--- a/test/src/unit-inspection.cpp
+++ b/test/src/unit-inspection.cpp
@@ -316,8 +316,8 @@ TEST_CASE("object inspection")
     SECTION("round trips")
     {
         for (const auto& s :
-                {"3.141592653589793", "1000000000000000010E5"
-                })
+    {"3.141592653589793", "1000000000000000010E5"
+    })
         {
             json j1 = json::parse(s);
             std::string s1 = j1.dump();

--- a/test/src/unit-udt.cpp
+++ b/test/src/unit-udt.cpp
@@ -711,3 +711,25 @@ TEST_CASE("an incomplete type does not trigger a compiler error in non-evaluated
 {
     static_assert(not is_constructible_patched<json, incomplete>::value, "");
 }
+
+namespace
+{
+class Evil
+{
+  public:
+    Evil() = default;
+    template <typename T>
+    Evil(T) {}
+};
+
+void from_json(const json&, Evil&) {}
+}
+
+TEST_CASE("Issue #924")
+{
+    // Prevent get<std::vector<Evil>>() to throw
+    auto j = json::array();
+
+    (void) j.get<Evil>();
+    (void) j.get<std::vector<Evil>>();
+}

